### PR TITLE
Update the img2mml service

### PR DESCRIPTION
**Pull request by ualiangzhang**
Update the image2mml service, fix bugs, improve the prediction performance

### CODE REVIEW

**Review by myedibleenso — CHANGES_REQUESTED** [review]
Hi, @ualiangzhang .  Thanks for tagging me.  There are a few things I noticed (some related to CI and others for clarity).


## Updating references to model path

- [ ] Model URL for tests (GitHub Actions): https://github.com/ml4ai/skema/blob/c1c82ae5c4582719b13bc4b8affaca9644a6b56a/.github/workflows/docs.yml#L58

- [ ] Model URL for tests (Drone CI; not currently active): https://github.com/ml4ai/skema/blob/c1c82ae5c4582719b13bc4b8affaca9644a6b56a/.drone.yml#L66

- [ ] img2mml/README.md: https://github.com/ml4ai/skema/blob/c1c82ae5c4582719b13bc4b8affaca9644a6b56a/skema/img2mml/README.md

## Path construction

There are few places where we could avoid hardcoding path separators:

  - [x] https://github.com/ml4ai/skema/pull/302/files#diff-802b106a9aa84777a13cf399ce033a63998ed037c9f3e8601f27ab67dd2ccb77R23 
  - [ ] https://github.com/ml4ai/skema/pull/302/files#diff-802b106a9aa84777a13cf399ce033a63998ed037c9f3e8601f27ab67dd2ccb77R23
  - ex `cwd / "trained_models/arxiv_im2mml_with_fonts_with_boldface_vocab.txt"` -> `cwd / "trained_models" / arxiv_im2mml_with_fonts_with_boldface_vocab.txt"`

## Questions
- Can you explain what this represents and why it is hard-coded? https://github.com/ml4ai/skema/pull/302/files#diff-0da916abe478c489b4818183fe0fb6fe787f333e329f9cefab0986a45e4c40c8R117

## [Recommended] Document steps needed to update repo when adding a new model

To simplify the process in the future, it would be helpful to document the steps needed to update the model (see https://github.com/ml4ai/skema/blob/c1c82ae5c4582719b13bc4b8affaca9644a6b56a/docs/dev/eqn2mml.md)

It would be good to note necessary changes to `Dockerfile.skema-py`, `.github/workflows/docs.yml`, etc.

### Dev instructions

We have some tips for configuring a local dev environment.  For completeness, it would be good to add something about retrieving the model to those instructions:

- [Add `RUN curl -L https://artifacts.askem.lum.ai/skema/img2mml/models/cnn_xfmer_arxiv_im2mml_with_fonts_boldface_best.pt > skema/img2mml/trained_models/cnn_xfmer_arxiv_im2mml_with_fonts_boldface_best.pt
RUN tree /app`](https://github.com/ml4ai/skema/blob/update_img2mml_service/docs/dev/env.md) before or after line related to fortran grammar


## [issue] Load or retrieve model

This is not a change I would suggest for this PR, but I think we might want to consider attempting to download the model if it isn't found.

Something like the following might work:

```python
try:
    import importlib.resources as pkg_resources
except ImportError:
    # Try backported to PY<37 `importlib_resources`.
    import importlib_resources as pkg_resources

MODEL_BASE_ADDRESS = "https://artifacts.lum.ai/skema/img2mml/models"
MODEL_NAME = "???"

from skema.img2mml import trained_models

def retrieve_model():
  # 1. check if model_file exists
  try:
      model_file = (pkg_resources.files(trained_models) / MODEL_NAME)
  # 2. if it doesn't exist, download to loc
  except:
    url = f"{MODEL_BASE_ADDRESS}/{MODEL_NAME}"
    pass
  # return model file location?
```

Shall we open an issue for that?

**Comment by myedibleenso** [comment]
@ualiangzhang , can you explain what this represents? https://github.com/ml4ai/skema/pull/302/files#diff-0da916abe478c489b4818183fe0fb6fe787f333e329f9cefab0986a45e4c40c8R117

File: skema/img2mml/api.py [hunk 1]
```diff
@@ -1,31 +1,45 @@
 import json
 import os
-from pathlib import Path
 import requests
-import re
-from skema.rest.proxies import SKEMA_MATHJAX_ADDRESS
+from pathlib import Path
+
+# from skema.rest.proxies import SKEMA_MATHJAX_ADDRESS
```
Comment (myedibleenso) [inline skema/img2mml/api.py:6]: You reference this later in your file, but you've commented out the import.

**Review by myedibleenso — CHANGES_REQUESTED** [review]
Thanks, @ualiangzhang .

Maybe I missed it, but I didn't see a unit test.  It would be good to at least ensure a) the img2mml model is found/retrieved, b) the model loads, and c) the model can be used to make predictions given some input.  The last test (making predictions) doesn't necessarily need to check the accuracy of the prediction (I assume you have benchmark tests in another repo for model validation); minimally we just want to confirm we get predictions.

File: skema/img2mml/api.py (hunk 1 above)
Comment (adarshp) [reply skema/img2mml/api.py]: @myedibleenso @ualiangzhang  - there was a small test here: https://github.com/ml4ai/skema/blob/main/skema/img2mml/tests/test_im2mml.py

But I don't know if it works anymore with @ualiangzhang 's changes. Also, it doesn't test the REST API, just the library function.

Comment (ualiangzhang) [reply skema/img2mml/api.py]: Thanks for clarification, @adarshp! I have been actually editing this file and reorganizing the code to test more.

**Review by myedibleenso — APPROVED** [review]

**Comment by myedibleenso** [comment]
Thank you, @ualiangzhang !  Great work!

**Pull request merged**